packages/hardhat-core/src/types/builtin-tasks/compile.ts

```diff
@@ -64,18 +64,21 @@ export interface DependencyGraph {
  */
 export interface CompilationJobsCreationResult {
   jobs: CompilationJob[];
-  errors: CompilationJobsCreationErrors;
+  errors: CompilationJobCreationError[];
 }
 
-export type CompilationJobsCreationErrors = {
-  [error in CompilationJobCreationError]?: string[];
-};
+export interface CompilationJobCreationError {
+  reason: CompilationJobCreationErrorReason;
+  file: ResolvedFile;
+  extra?: any;
+}
```

Comment:
This is the main type change. The `CompilationJobsCreationError` type, that was a map from an error tipe to a list of source names with that error, is gone. Instead I used a list of errors. Each error has a reason (this is the extendable enum previously called CompilationJobCreationError), a file, and an optional "extra" type, that can have more information about the error.

I don't love having an `any` for the extra field, but I'm not sure it's worth it to do something trickier here.